torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -497,26 +507,46 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
         futureTimeouts_.begin()->first == INFINITE_TIMEOUT) {
       sleepTime = INFINITE_TIMEOUT;
     } else {
-      const auto minFutureExpirationTime = futureTimeouts_.begin()->first;
-      const auto remainingTime = getRPCRemainingTime(minFutureExpirationTime);
+      knownEarliestEndTime = futureTimeouts_.begin()->first;
+      const auto remainingTime = getRPCRemainingTime(knownEarliestEndTime);
       sleepTime = std::max(remainingTime, std::chrono::milliseconds(0));
     }
 
+    bool shouldUpdatesleepTime = true;
     if (sleepTime == INFINITE_TIMEOUT) {
-      futureTimeoutCV_.wait(lock);
+      futureTimeoutCV_.wait(lock, [&, this]() {
+        // Do not wait on the cond var if RPC has shutdown.
+        // Notice, whoever modifying `shutdown_`
+        // must acquire lock on `futureMutex_`.
+        if (shutdown_.load()) {
+          return true;
+        }
+        auto earliestEndTimeInMap = futureTimeouts_.begin()->first;
+        return earliestEndTimeInMap < knownEarliestEndTime;
+      });
     } else {
-      futureTimeoutCV_.wait_for(lock, sleepTime);
+      shouldUpdatesleepTime =
+          futureTimeoutCV_.wait_for(lock, sleepTime, [&, this]() {
+            // Do not wait on the cond var if RPC has shutdown.
+            // Notice, whoever modifying `shutdown_`
+            // must acquire lock on `futureMutex_`.
+            if (shutdown_.load()) {
+              return true;
+            }
+            auto earliestEndTimeInMap = futureTimeouts_.begin()->first;
+            return earliestEndTimeInMap < knownEarliestEndTime;
```

Comment:
This is duplication to the lambda in the if clause. Let's consolidate them.

Reply:
outdated. does not exist now.